File: 1406-StoneGameIII/1406-StoneGameIII.cpp
// Last updated: 8/3/2026, 5:17:27 PM
1class Solution {
2public:
3    string stoneGameIII(vector<int>& stoneValue) 
4    {
5       int n = stoneValue.size(); 
6       vector<int> dp(3, 0); 
7
8       for (int i = n - 1; i >= 0; i--) {
9            int takeOne = stoneValue[i] - dp[(i + 1) % 3];
10
11            int takeTwo = INT_MIN;
12            if (i + 1 < n) takeTwo = stoneValue[i] + stoneValue[i + 1] - dp[(i + 2) % 3];
13
14            int takeThree = INT_MIN;
15            if (i + 2 < n) takeThree = stoneValue[i] + stoneValue[i + 1] + stoneValue[i + 2] - dp[(i + 3) % 3];
16
17            dp[i % 3] = max({takeOne, takeTwo, takeThree});
18        } 
19
20       int value = dp[0]; 
21       if(value > 0)
22           return "Alice";
23        else if(value  < 0)
24            return "Bob";
25        else
26            return "Tie";
27    }
28};

File: 1406-StoneGameIii/1406-StoneGameIii.py
# Last updated: 9/5/2026, 12:31:06 AM
class Solution:
    def stoneGameIII(self, st: List[int]) -> str:
        n = len(st)
        dp = [[0] * (n + 1) for _ in range(2)]

        for i in range(n - 1, -1, -1):
            for per in range(2):
                stones = 0
                dp[1][i] = float('-inf')
                dp[0][i] = float('inf')

                for x in range(1, min(n - i + 1, 4)):
                    stones += st[i + x - 1]
                    dp[1][i] = max(dp[1][i], stones + dp[0][i + x])
                    dp[0][i] = min(dp[0][i], -stones + dp[1][i + x])

        ans = dp[1][0]
        return "Tie" if ans == 0 else ("Alice" if ans > 0 else "Bob")
